core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleMatchVisualizer.java

```diff
@@ -0,0 +1,300 @@
+ *   new VolcanoRuleMatchVisualizerListener(volcanoPlanner);
+ * volcanoPlanner.addListener(visualizerListener);
+ *
+ * volcanoPlanner.findBestExpr();
+ *
+ * // after the optimization, adds the final best plan
+ * visualizerListener.getVisualizer().addFinalPlan();
+ * // writes the output to files
+ * visualizerListener.getVisualizer().writeToFile(outputDirectory, "");
+ * </pre>
+ */
+public class VolcanoRuleMatchVisualizer {
+
+  VolcanoPlanner volcanoPlanner;
+
+  // a sequence of ruleMatch ID to represent the order of rule match
+  List<String> ruleMatchSequence = new ArrayList<>();
+  // map of ruleMatch ID and the info, including the state snapshot at the time of ruleMatch
+  Map<String, VisualizerRuleMatchInfo> ruleInfoMap = new HashMap<>();
+  // map of nodeID to the ruleID it's first added
+  Map<String, String> nodeAddedInRule = new HashMap<>();
+
+  // a map of relNode ID to the actual RelNode object
+  // contains all the relNodes appear during the optimization
+  // all RelNode are immutable in Calcite, therefore only new nodes will be added
+  Map<String, RelNode> allNodes = new HashMap<>();
+
+  public VolcanoRuleMatchVisualizer(VolcanoPlanner volcanoPlanner) {
+    this.volcanoPlanner = volcanoPlanner;
+  }
+
+  public void addRuleMatch(String ruleCallID, Collection<? extends RelNode> matchedRels) {
+
+    // store the current state snapshot
+    // nodes contained in the sets
+    // and inputs of relNodes (and relSubsets)
+    Map<String, String> setLabels = new HashMap<>();
+    Map<String, String> setOriginalRel = new HashMap<>();
+    Map<String, Set<String>> nodesInSet = new HashMap<>();
+    Map<String, Set<String>> nodeInputs = new HashMap<>();
+
+    // newNodes appeared after this ruleCall
+    Set<String> newNodes = new HashSet<>();
+
+    // populate current snapshot, and fill in the allNodes map
+    volcanoPlanner.allSets.forEach(set -> {
+      String setID = "set-" + set.id;
+      String setLabel = getSetLabel(set);
+      setLabels.put(setID, setLabel);
+      setOriginalRel.put(setID, set.rel == null ? "" : String.valueOf(set.rel.getId()));
+
+      nodesInSet.put(setID, nodesInSet.getOrDefault(setID, new HashSet<>()));
+
+      Consumer<RelNode> addNode = rel -> {
+        String nodeID = String.valueOf(rel.getId());
+        nodesInSet.get(setID).add(nodeID);
+
+        if (!allNodes.containsKey(nodeID)) {
+          newNodes.add(nodeID);
+          allNodes.put(nodeID, rel);
+        }
+      };
+
+      Consumer<RelNode> addLink = rel -> {
+        String nodeID = String.valueOf(rel.getId());
+        nodeInputs.put(nodeID, new HashSet<>());
+        if (rel instanceof RelSubset) {
+          RelSubset relSubset = (RelSubset) rel;
+          relSubset.getRelList().stream()
+              .filter(input -> input.getTraitSet().equals(relSubset.getTraitSet()))
+              .forEach(input -> nodeInputs.get(nodeID).add(String.valueOf(input.getId())));
```

Comment:
This was a design tradeoff to make the links in a set less messy. In the environment I was working on, a set contains multiple subsets with related traits. So a relNode belongs to many subsets at the same time. 

For example, supposethere are 2 subsets `S1`, `S2` (`S2` satisfies `S1`) and 2 relnodes `R1`, `R2` with the same trait as `S2`. 

Instead of creating 4 links for relNodes: `R1 -> S1`, `R1 -> S2`, `R2 -> S1`, `R2 -> S2`,
I decided to draw 2 links for relNodes: `R1 -> S2` and `R2 -> S2`, and 1 indirect link for the subsets `S2 -> S1`. So the subset satisfaction relationship is indirectly shown. 

This way in a set with many subsets/relnodes, the links are much less messy.

Reply:
Thanks for the clarification. It's better to reduce the number of arrows in the graph, so let's keep it as it is.